--- a/autoload/getopts/cache.vim
+++ b/autoload/getopts/cache.vim
@@ -5,11 +5,18 @@
 let s:cache_path = fnamemodify(s:scr, ':p:h')

 function! getopts#cache#getcachedopts(ext, optfunc)
-  let s:cache = s:cache_path . '/' . &filetype . '.' . a:ext . '.cache'
-  if !s:CacheExists()
-    call s:CreateCache(a:optfunc)
+  if !exists('g:clang_no_cache')
+    let g:clang_no_cache = 0
+  endif
+  if g:clang_no_cache == 0
+    let s:cache = s:cache_path . '/' . &filetype . '.' . a:ext . '.cache'
+    if !s:CacheExists()
+      call s:CreateCache(a:optfunc)
+    endif
+    call s:ReadCache()
+  else
+    let s:opts = s:GetOptions(a:optfunc)
   endif
-  call s:ReadCache()
   call s:AppendOptions()
 endfunction

