--- a/autoload/getopts/gcc.vim
+++ b/autoload/getopts/gcc.vim
@@ -6,10 +6,17 @@

 function! getopts#gcc#getopts()
   call s:DetemineFileType()
-  if !s:CacheExists()
-    call s:CreateCache()
+  if !exists('g:clang_no_cache')
+    let g:clang_no_cache = 0
+  endif
+  if g:clang_no_cache == 0
+    if !s:CacheExists()
+      call s:CreateCache()
+    endif
+    call s:ReadCache()
+  else
+    let b:gcc_opts = s:GetGCCOptions()
   endif
-  call s:ReadCache()
   call s:AppendOptions()
 endfunction

