--- a/mcabber/xmpp.c
+++ b/mcabber/xmpp.c
@@ -1621,7 +1621,9 @@
 {
   char *r;
   char *buf;
+  GSList* roster_node;
   int newbuddy;
+  guint subscription;
   guint hook_result;
   LmMessageSubType mstype = lm_message_get_sub_type(m);
   const char *from = lm_message_get_from(m);
@@ -1639,17 +1641,22 @@
   }
   r = jidtodisp(from);

-  newbuddy = !roster_find(r, jidsearch, 0);
+  roster_node = roster_find(r, jidsearch, 0);
+  if (roster_node) {
+    newbuddy = 0;
+    subscription = buddy_getsubscription(roster_node->data);
+  } else {
+    newbuddy = 1;
+    subscription = 0;
+  }

   hook_result = hk_subscription(mstype, r, msg);

   if (mstype == LM_MESSAGE_SUB_TYPE_SUBSCRIBE) {
     /* The sender wishes to subscribe to our presence */

-    if (hook_result) {
-      g_free(r);
-      return LM_HANDLER_RESULT_REMOVE_MESSAGE;
-    }
+    if (hook_result || (subscription & sub_from))
+      goto handle_s10n_return;

     buf = g_strdup_printf("<%s> wants to subscribe to your presence updates",
                           from);
@@ -1685,6 +1692,8 @@
   } else if (mstype == LM_MESSAGE_SUB_TYPE_UNSUBSCRIBE) {
     /* The sender is unsubscribing from our presence */
     xmpp_send_s10n(from, LM_MESSAGE_SUB_TYPE_UNSUBSCRIBED);
+    if (!(subscription & sub_from))
+      goto handle_s10n_return;
     buf = g_strdup_printf("<%s> is unsubscribing from your "
                           "presence updates", from);
     scr_WriteIncomingMessage(r, buf, 0, HBB_PREFIX_INFO, 0);
@@ -1692,6 +1701,8 @@
     g_free(buf);
   } else if (mstype == LM_MESSAGE_SUB_TYPE_SUBSCRIBED) {
     /* The sender has allowed us to receive their presence */
+    if (subscription & sub_to)
+      goto handle_s10n_return;
     buf = g_strdup_printf("<%s> has allowed you to receive their "
                           "presence updates", from);
     scr_WriteIncomingMessage(r, buf, 0, HBB_PREFIX_INFO, 0);
@@ -1702,6 +1713,8 @@
        subscription has been cancelled */
     roster_unsubscribed(from);
     scr_update_roster();
+    if (!(subscription & sub_to))
+      goto handle_s10n_return;
     buf = g_strdup_printf("<%s> has cancelled your subscription to "
                           "their presence updates", from);
     scr_WriteIncomingMessage(r, buf, 0, HBB_PREFIX_INFO, 0);
@@ -1714,6 +1727,8 @@

   if (newbuddy)
     scr_update_roster();
+
+handle_s10n_return:
   g_free(r);
   return LM_HANDLER_RESULT_REMOVE_MESSAGE;
 }
