--- a/mcabber/compl.c
+++ b/mcabber/compl.c
@@ -159,11 +159,11 @@
           compval = g_strdup_printf("%s%s", word+len, suffix);
         else
           compval = g_strdup(word+len);
-        c->list = g_slist_insert_sorted(c->list, compval,
-                                        (GCompareFunc)g_ascii_strcasecmp);
+        c->list = g_slist_prepend(c->list, compval);
       }
     }
   }
+  c->list = g_slist_reverse(c->list);
   c->next = c->list;
   InputCompl = c;
   return g_slist_length(c->list);
@@ -218,39 +218,41 @@
   return r;
 }

-
 /* Categories functions */

-//  compl_add_category_word(categ, command)
-// Adds a keyword as a possible completion in category categ.
-void compl_add_category_word(guint categ, const gchar *word)
+static category* prepare_category(guint categ, _Bool create)
 {
   guint64 catv;
   GSList *sl_cat;
-  category *cat;
-  char *nword;
+  category *cat = 0;

-  if (!categ) {
+  if (categ) {
+    categ--;
+    catv = 1UL << categ;
+
+    // Look for category
+    for (sl_cat=Categories; sl_cat; sl_cat = g_slist_next(sl_cat)) {
+      if (catv == ((category*)sl_cat->data)->flag)
+        break;
+    }
+    if (!sl_cat) {   // Category not found, let's create it
+      if (create) {
+        cat = g_new0(category, 1);
+        cat->flag = catv;
+        Categories = g_slist_append(Categories, cat);
+      }
+    } else
+      cat = (category*)sl_cat->data;
+  } else {
     scr_log_print(LPRINT_LOGNORM, "Error: compl_add_category_word() - "
                   "Invalid category.");
-    return;
   }
+  return cat;
+}

-  categ--;
-  catv = 1UL << categ;
-
-  // Look for category
-  for (sl_cat=Categories; sl_cat; sl_cat = g_slist_next(sl_cat)) {
-    if (catv == ((category*)sl_cat->data)->flag)
-      break;
-  }
-  if (!sl_cat) {   // Category not found, let's create it
-    cat = g_new0(category, 1);
-    cat->flag = catv;
-    Categories = g_slist_append(Categories, cat);
-  } else
-    cat = (category*)sl_cat->data;
-
+static gchar* prepare_word(const gchar* word)
+{
+  gchar* nword;
   // If word is not space-terminated, we add one trailing space
   for (nword = (char*)word; *nword; nword++)
     ;
@@ -260,59 +262,69 @@
   } else {              // word is fine
     nword = g_strdup(word);
   }
+  return nword;
+}

-  if (g_slist_find_custom(cat->words, nword, (GCompareFunc)g_strcmp0) != NULL)
-    return;
+//  compl_add_category_word(categ, command)
+// Adds a keyword as a possible completion in category categ.
+void compl_add_category_word(guint categ, const gchar *word)
+{
+  gchar *nword;
+  category *cat = prepare_category(categ, 1);

-  cat->words = g_slist_insert_sorted(cat->words, nword,
-                                     (GCompareFunc)g_ascii_strcasecmp);
+  if (cat) {
+    nword = prepare_word(word);
+
+    if (g_slist_find_custom(cat->words, nword, (GCompareFunc)g_strcmp0) != NULL) {
+      g_free(nword);
+      return;
+    }
+
+    cat->words = g_slist_insert_sorted(cat->words, nword,
+                                       (GCompareFunc)g_ascii_strcasecmp);
+  }
 }

-//  compl_del_category_word(categ, command)
-// Removes a keyword from category categ in completion list.
-void compl_del_category_word(guint categ, const gchar *word)
+//  compl_add_category_word(categ, command)
+// Adds a keyword as a possible completion in category categ.
+// New keyword will be put at the begginning of the category keywords list.
+void compl_add_unordered_category_word(guint categ, const gchar *word)
 {
-  guint64 catv;
-  GSList *sl_cat, *sl_elt;
-  category *cat;
-  char *nword;
+  gchar *nword;
+  category *cat = prepare_category(categ, 1);

-  if (!categ) {
-    scr_log_print(LPRINT_LOGNORM, "Error: compl_del_category_word() - "
-                  "Invalid category.");
-    return;
-  }
+  if (cat) {
+    nword = prepare_word(word);

-  categ--;
-  catv = 1UL << categ;
+    if (g_slist_find_custom(cat->words, nword, (GCompareFunc)g_strcmp0) != NULL) {
+      g_free(nword);
+      return;
+    }

-  // Look for category
-  for (sl_cat=Categories; sl_cat; sl_cat = g_slist_next(sl_cat)) {
-    if (catv == ((category*)sl_cat->data)->flag)
-      break;
+    cat->words = g_slist_prepend(cat->words, nword);
   }
-  if (!sl_cat) return;   // Category not found, finished!
+}

-  cat = (category*)sl_cat->data;
+//  compl_del_category_word(categ, command)
+// Removes a keyword from category categ in completion list.
+void compl_del_category_word(guint categ, const gchar *word)
+{
+  GSList *sl_elt;
+  char *nword;
+  category *cat = prepare_category(categ, 0);

-  // If word is not space-terminated, we add one trailing space
-  for (nword = (char*)word; *nword; nword++)
-    ;
-  if (nword > word) nword--;
-  if (*nword != ' ') {  // Add a space
-    nword = g_strdup_printf("%s ", word);
-  } else {              // word is fine
-    nword = g_strdup(word);
-  }
+  if (cat) {
+    nword = prepare_word(word);

-  sl_elt = cat->words;
-  while (sl_elt) {
-    if (!strcasecmp((char*)sl_elt->data, nword)) {
-      g_free(sl_elt->data);
-      cat->words = g_slist_delete_link(cat->words, sl_elt);
-      break; // Only remove first occurence
+    sl_elt = cat->words;
+    while (sl_elt) {
+      if (!strcasecmp((char*)sl_elt->data, nword)) {
+        g_free(sl_elt->data);
+        cat->words = g_slist_delete_link(cat->words, sl_elt);
+        break; // Only remove first occurence
+      }
+      sl_elt = g_slist_next(sl_elt);
     }
-    sl_elt = g_slist_next(sl_elt);
   }
 }

--- a/mcabber/compl.h
+++ b/mcabber/compl.h
@@ -35,6 +35,7 @@
 #endif

 void    compl_add_category_word(guint categ, const gchar *command);
+void    compl_add_unordered_category_word(guint categ, const gchar *command);
 void    compl_del_category_word(guint categ, const gchar *word);
 GSList *compl_get_category_list(guint categ, guint *dynlist);
