diff -ur screen-4.0.3.orig//display.h screen-4.0.3/display.h
--- screen-4.0.3.orig//display.h	2003-07-01 16:01:42.000000000 +0200
+++ screen-4.0.3/display.h	2010-11-17 13:18:13.182984179 +0100
@@ -22,6 +22,16 @@
  * $Id: 4.0.3-extend-d_termname.patch,v 1.1 2010/11/17 14:51:18 jlec Exp $ FAU
  */
 
+#include <limits.h>
+
+#ifndef NAME_MAX
+# ifndef MAXNAMELEN
+#  define NAME_MAX 255
+# else
+#  define NAME_MAX MAXNAMELEN
+# endif
+#endif
+
 #ifdef MAPKEYS
 
 #define KMAP_KEYS (T_OCAPS-T_CAPS)
@@ -85,7 +95,7 @@
   struct win *d_other;		/* pointer to other window */
   int   d_nonblock;		/* -1 don't block if obufmax reached */
 				/* >0: block after nonblock secs */
-  char  d_termname[20 + 1];	/* $TERM */
+  char  d_termname[NAME_MAX + 1]; /* $TERM */
   char	*d_tentry;		/* buffer for tgetstr */
   char	d_tcinited;		/* termcap inited flag */
   int	d_width, d_height;	/* width/height of the screen */
