--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -29,12 +29,6 @@
 set(AALIB_LIBRARY_DIRS "/usr/lib" CACHE FILEPATH "Path, where AAlib is located")
 set(AALIB_INCLUDE_DIRS "/usr/include" CACHE FILEPATH "Path to AAlib includes")
 set(AALIB_LIBRARIES "-laa" CACHE TEXT "Libraries, necessary to link with AAlib")
-include(CheckIncludeFile)
-set(CMAKE_REQUIRED_INCLUDES ${MCABBER_INCLUDE_DIRS})
-check_include_file(mcabber/pep.h HAVE_MCABBER_PEP_H)
-if(NOT HAVE_MCABBER_PEP_H)
-	message(SEND_ERROR "Pep module header not found")
-endif()
 include(CheckSymbolExists)
 set(CMAKE_REQUIRED_INCLUDES ${LM_INCLUDE_DIRS})
 set(CMAKE_REQUIRED_LIBRARIES ${LM_LIBRARIES})
